include/adt/hash.h

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2021 Scott Vokes
+ *
+ * See LICENCE for the full copyright terms.
+ */
+
+#ifndef FSM_HASH_H
+#define FSM_HASH_H
+
+#include <stdint.h>
+
+#include "common/check.h"
+
+/* 32 and 64-bit approximations of the golden ratio. */
+#define FSM_PHI_32 0x9e3779b9UL
+#define FSM_PHI_64 (uint64_t)0x9e3779b97f4a7c15UL
+
+/* A suitable hash function for individual sequentially allocated
+ * identifiers. See Knuth 6.4, Fibonacci hashing. */
+
+SUPPRESS_EXPECTED_UNSIGNED_INTEGER_OVERFLOW()
+static __inline__ uint64_t
+fsm_hash_id(unsigned id)
+{
+	return FSM_PHI_64 * (uint64_t)(id + (unsigned)1);
+}
+
+/* FNV-1a hash function, 32 and 64 bit versions. This is in the public
+ * domain. For details, see:
+ *
+ *     http://www.isthe.com/chongo/tech/comp/fnv/index.html
+ */
+
+SUPPRESS_EXPECTED_UNSIGNED_INTEGER_OVERFLOW()
+static __inline__ uint32_t
+fsm_hash_fnv1a_32(const uint8_t *buf, size_t length)
```

Comment:
The `fsm_` prefix is just for symbols with extern linkage (ideally what's only in the public API); these would be just `hash_fnv1a_32()` here

Reply:
Changed in e68fdc5e.